### PR TITLE
Add single dimension -> multidimension partition mapping

### PR DESCRIPTION
**Pull request by clairelin135**
Resolves https://github.com/dagster-io/dagster/issues/10722

Following https://github.com/dagster-io/dagster/pull/10909/files, this PR adds a partition mapping from an upstream single-dimensionally partitioned asset to a downstream multipartitioned asset. In order to use this class, the upstream partitions definition must be a dimension of the downstream asset.

The new `PartitionMapping` object also returns partition subsets instead of ranges, enabling use cases where a downstream subset is not a valid range e.g.:
```
MultiPartitionsDefinition(
    {"ab": StaticPartitionsDefinition(["a", "b"]), "12": StaticPartitionsDefinition(["1", "2"])}
)
```
has partition keys: ['1|a', '1|b', '2|a', '2|b']
If the upstream asset had partitions definition StaticPartitionsDefinition(["a", "b"]), the downstream partition keys subset should be {'1|a', '2|a'}, which can't be expressed as a range.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/partition_mapping.py [hunk 1]
```diff
@@ -193,6 +197,145 @@ def get_downstream_partitions_for_partition_range(
         raise NotImplementedError()
 
 
+@experimental
+class SingleDimensionToMultiPartitionMapping(PartitionMapping):
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/partition_mapping.py:201]: From the name, I wasn't able to tell which direction this goes - i.e. whether the upstream or the downstream partitions definition is expected to be multi-. Is there a way to make that more clear, or at least conform to a convention?

FWIW `AllPartitionMapping` and `LastPartitionMapping`, prefer to what partitions in the upstream asset correspond to a partition of the downstream asset.

Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/partition_mapping.py:227]: That's a good point. Renaming this to `SingleDimensionDependencyMapping`, hopefully that is clearer.

Tossing out some possible naming ideas:
- SingleDimensionDependencyMapping
- MatchesDownstreamPartitionDimensionMapping
- PartitionDimensionMapping

**Comment by clairelin135** [comment]
@sryza per your feedback I've renamed the class and switched to the new partition subset methods, raising `NotImplementedError`s for the range methods. I think this is ready for you to take another look!

File: python_modules/dagster/dagster/_core/definitions/partition_mapping.py [hunk 2]
```diff
@@ -215,6 +219,107 @@ def get_downstream_partitions_for_partition_range(
...
+        matching_dimensions = [
+            partitions_def
+            for partitions_def in downstream_partitions_def.partitions_defs
+            if partitions_def.name == self.partition_dimension_name
+        ]
+        if len(matching_dimensions) != 1:
+            check.failed(f"Partition dimension '{self.partition_dimension_name}' not found")
+        matching_dimension_def = next(iter(matching_dimensions))
+
+        if upstream_partitions_def != matching_dimension_def.partitions_def:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/partition_mapping.py:254]: Something to keep in mind for future optimization is that this `!=` could be expensive for large `StaticPartitionsDefinition`s.

**Review by sryza — APPROVED** [review]
Nice

**Pull request merged**